### PR TITLE
Framework: Enhance find_package(Trilinos) test

### PR DESCRIPTION
**Pull request by sebrowne**
Enhance find_package(Trilinos) test to catch issues like #12371
@trilinos/framework

## Motivation
Turning on LINK_LIBRARIES_ONLY_TARGETS caused an issue with a downstream customer when we did not set up the installation correctly with respect to imported targets.

## Related Issues
#12362 
#12371 
#10081 

## Testing
Tested with the changes from #12362 and it immediately caught the issue.

### CODE REVIEW

File: packages/TrilinosInstallTests/find_package_Trilinos/CMakeLists.txt
Comment (bartlettroscoe) [inline packages/TrilinosInstallTests/find_package_Trilinos/CMakeLists.txt:13]: My only suggestion to make this more portable is to perhaps changes to the a `hello.cpp` C++ file so that the C++ linker will be used.  On some systems, linking with a C linker involving C++ libraries will cause an error.

**Review by bartlettroscoe — COMMENTED** [review]
It is excellent this caught the error produced by PR:

* #12362

But it might improve portability to make this a C++ dummy executable using the C++ compiler as the linker.

**Comment by bartlettroscoe** [comment]
@sebrowne, it is possible that adding a build step this may catch additional issues.  For example, if a library file does not exist on the system, then that will only be caught at build time, not configure time.

Can we consider adding the additional build test step:

```
  TEST_2
    MESSAGE "Build dummy project that calls find_package(Trilinos)"
    CMND "${CMAKE_COMMAND}"
    ARGS --build .
```

to this `tribits_add_advanced_test(find_package_Trilinos ...)` test right after the `TEST_0` block at:

https://github.com/trilinos/Trilinos/blob/cb4090def796ab4ecdc5aa099329c4b8534d40a5/packages/TrilinosInstallTests/CMakeLists.txt#L256

?

**Comment by sebrowne** [comment]
@bartlettroscoe yes, not a problem.  I didn't look closely enough at the test to realize that it's not doing that, so adding that seems great to me.

**Comment by sebrowne** [comment]
@bartlettroscoe do I need `ALWAYS_FAIL_ON_NONZERO_RETURN`?  Or is that only because of the regex match for the configure block

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe do I need `ALWAYS_FAIL_ON_NONZERO_RETURN`? Or is that only because of the regex match for the configure block

Only if you want to assert that the return code is 0 **in addition** to greping the STDOUT. See the first set of If/Else statements in:

* https://tribitspub.github.io/TriBITS/users_guide/index.html#test-case-pass-fail-tribits-add-advanced-test

where the default pass/fail criteria is just:

```
  If command return code == 0:
    TEST_CASE_PASSED = TRUE
  Endif
```

NOTE: Raw CMake/CTes does **not** allow you to check **both** the STDOUT **and** the return code.  (I have seem major defects skip through over the years by not checking both.)

**Comment by bartlettroscoe** [comment]
We see the updated test `TrilinosInstallTests_find_package_Trilinos` passing in the last PR testing iteration in [this query](https://trilinos-cdash.sandia.gov/queryTests.php?project=Trilinos&date=2023-10-10&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=PR-12389-&field2=testname&compare2=61&value2=TrilinosInstallTests_find_package_Trilinos) showing:

<html>
<body>


Site | Build Name | Test Name | Status | Time Status | Time | Proc Time | Details | Build Time | Processors
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
ascic147 | PR-12389-test-rhel7_sems-intel-2021.3-sems-openmpi-4.0.5_release-debug_shared_no-kokkos-arch_no-asan_no-complex_fpic_mpi_no-pt_no-rdc_no-uvm_deprecated-off_no-package-enables-1208 | TrilinosInstallTests_find_package_Trilinos | Passed | Passed | 2s 920ms | 2s 920ms | Completed | 2023-10-10T19:35:18 MDT | 1
ascic142 | PR-12389-test-rhel7_sems-clang-11.0.1-openmpi-1.10.1-serial_release-debug_shared_no-kokkos-arch_no-asan_no-complex_no-fpic_mpi_no-pt_no-rdc_no-uvm_deprecated-on_no-package-enables-1566 | TrilinosInstallTests_find_package_Trilinos | Passed | Passed | 1s 930ms | 1s 930ms | Completed | 2023-10-10T19:34:09 MDT | 1
ascicgpu050 | PR-12389-test-rhel7_sems-cuda-11.4.2-sems-gnu-10.1.0-sems-openmpi-4.0.5_release_static_Volta70_no-asan_complex_no-fpic_mpi_pt_no-rdc_no-uvm_deprecated-on_no-package-enables-2569 | TrilinosInstallTests_find_package_Trilinos | Passed | Passed | 5s 220ms | 5s 220ms | Completed | 2023-10-10T19:34:08 MDT | 1
ascic114 | PR-12389-test-rhel7_sems-gnu-8.3.0-openmpi-1.10.1-openmp_release-debug_static_no-kokkos-arch_no-asan_no-complex_no-fpic_mpi_no-pt_no-rdc_no-uvm_deprecated-on_no-package-enables-3078 | TrilinosInstallTests_find_package_Trilinos | Passed | Passed | 3s 120ms | 3s 120ms | Completed | 2023-10-10T19:33:56 MDT | 1
ascic166 | PR-12389-test-rhel7_sems-gnu-8.3.0-openmpi-1.10.1-serial_debug_shared_no-kokkos-arch_no-asan_no-complex_no-fpic_mpi_no-pt_no-rdc_no-uvm_deprecated-on_no-package-enables-1567 | TrilinosInstallTests_find_package_Trilinos | Passed | Passed | 1s 370ms | 1s 370ms | Completed | 2023-10-10T19:33:33 MDT | 1
ascic158 | PR-12389-test-rhel7_sems-v2-gnu-8.3.0-serial_release-debug_shared_no-kokkos-arch_no-asan_no-complex_no-fpic_no-mpi_no-pt_no-rdc_no-uvm_deprecated-on_no-package-enables-1577 | TrilinosInstallTests_find_package_Trilinos | Passed | Passed | 1s 620ms | 1s 620ms | Completed | 2023-10-10T19:33:28 MDT | 1


</body>
</html>

You can see that built target "Built target hello" in [this query](https://trilinos-cdash.sandia.gov/queryTests.php?project=Trilinos&date=2023-10-10&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=PR-12389-&field2=testname&compare2=61&value2=TrilinosInstallTests_find_package_Trilinos&field3=testoutput&compare3=95&value3=Built%20target%20hello) (click "Show Matching Output").

My only concern is that since only the `TrilinosInstallTests` package was changed, TriBITS only enabled that package and therefore this did not pull in any real libraries, only the dummy library `TrilinosInstallTests::all_libs` as shown in [this query](https://trilinos-cdash.sandia.gov/queryTests.php?project=Trilinos&date=2023-10-10&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=PR-12389-&field2=testname&compare2=61&value2=TrilinosInstallTests_find_package_Trilinos&field3=testoutput&compare3=95&value3=Trilinos_LIBRARIES) (click "Show Matching Output").  The hope is that this will also pass when linking in all of the Trilinos libraries, but we can't see that from these PR builds  & tests.  Therefore, it might be safer to make a trivial change to the base `CMakeLists.txt` file and trigger the enable of all Trilinos packages to make sure this also works.  But my guess is that this will be fine and is safe to merge.

@sebrowne, I will go ahead and approve the PR after taking off the `AT: AUTOMERGE` label, and if you are comfortable with this level of testing, just go ahead and merge the PR.

**Review by bartlettroscoe — DISMISSED** [review]
I will go ahead and approve this and leave it up to @sebrowne if he is comfortable with not enabling all of the Trilinos packages in these PR builds (see [above](https://github.com/trilinos/Trilinos/pull/12389#issuecomment-1758527589).

@sebrowne, if you are comfortable with this passing when all of the Trilinos packages are enabled (which you could test locally pretty fast), please go ahead and click "merge".

**Comment by sebrowne** [comment]
@bartlettroscoe safest to trigger a PR check against all packages.  I'm in no great rush to get this merged.

**Review by bartlettroscoe — APPROVED** [review]
With the triggering of the enable of all the packages and linking against all of the packages (see [above](https://github.com/trilinos/Trilinos/pull/12389#issuecomment-1771039532)), this is about a good of a test as you can gets for cheap native test of `find_package(Trilinos)`.

**Pull request merged**
